test/StaticAnalysis/narrow-some-none.php

```diff
@@ -0,0 +1,34 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Prewk\Option\Test\StaticAnalysis;
+
+use Prewk\Option;
+use Prewk\Option\None;
+use Prewk\Option\Some;
+
+function foo(Option $option): void
+{
+    if ($option->isSome()) {
+        bar($option);
+    }
+
+    if ($option->isNone()) {
+        baz($option);
+    }
+}
+
+/**
+ * @psalm-suppress UnusedParam
+ */
+function bar(Some $some): void
+{
+}
+
+/**
+ * @psalm-suppress UnusedParam
+ */
+function baz(None $none): void
+{
+}
```

Comment:
question: Can you TLDR how the static analysis stuff works? Does it help psalm somehow?

Reply:
It doesn't help psalm, it just checks that our type assertions are correct.

Here, we check eg. if after `Option#isSome()` returns `true`, psalm understands that we have a `Some` instance by passing it to a function that expects a `Some` instance as its first arg. Since psalm doesn't generate an error for this, it means our type assertions are valid.